ACPs/77-reinventing-subnets/README.md

```diff
@@ -32,48 +32,202 @@ With no Elastic Subnets live on Mainnet, it is clear that Permissionless Subnets
+| | | 8 + len(`addresses`) * 20 bytes |
+
+- `threshold` is the number of `addresses` that must provide a signature for the `PChainOwner` to authorize an action.
+- In order to be a valid:
+   - If `threshold` is `0`, `addresses` must be empty
+   - `threshold` <= len(`addresses`)
+   - Entries of `addresses` must be unique and sorted in ascending order
+
+The `RegisterSubnetValidatorMessage` is specified as an `AddressedCall` with a payload of:
+
+| Field | Type | Size |
+| -: | -: | -: |
+| `codecID` | `uint16` | 2 bytes |
+| `typeID` | `uint32` | 4 bytes |
+| `subnetID` | `[32]byte` | 32 bytes |
+| `nodeID` | `[]byte` | 4 + len(`nodeID`) bytes |
+| `blsPublicKey` | `[48]byte` | 48 bytes |
+| `expiry` | `uint64` | 8 bytes |
+| `remainingBalanceOwner` | `PChainOwner` | 8 + len(`addresses`) * 20 bytes |
+| `disableOwner` | `PChainOwner` | 8 + len(`addresses`) * 20 bytes |
+| `weight` | `uint64` | 8 bytes |
+| | | 122 + len(`nodeID`) + (len(`addresses1`) + len(`addresses2`)) * 20 bytes |
+
+- `codecID` is the codec version used to serialize the payload, and is hardcoded to `0x0000`
+- `typeID` is the payload type identifier and is `0x00000001` for this payload
+- `subnetID`, `nodeID`, `weight`, and `blsPublicKey` are for the Subnet Validator being added
+- `expiry` is the time at which this message becomes invalid. As of a P-Chain timestamp `>= expiry`, this Avalanche Warp Message can no longer be used to add the `nodeID` to the validator set of `subnetID`
+- `remainingBalanceOwner` is the P-Chain owner where leftover $AVAX from the Subnet Validator's Balance will be issued to when this validator it is removed from the validator set.
+- `disableOwner` is the only P-Chain owner allowed to disable the validator using `DisableSubnetValidatorTx`, specified below.
+
+#### `SubnetValidatorRegistrationMessage`
+
+A `SubnetValidatorRegistrationMessage` is sent from the P-Chain as an acknowledgement to inform validator managers when a validation period has either begun or been invalidated.
+
+The `SubnetValidatorRegistrationMessage` is specified as an `AddressedCall` with `sourceChainID` set to the P-Chain ID, the `sourceAddress` set to an empty byte array, and a payload of:
+
+| Field | Type | Size |
+| -: | -: | -: |
+| `codecID` | `uint16` | 2 bytes |
+| `typeID` | `uint32` | 4 bytes |
+| `validationID` | `[32]byte` | 32 bytes |
+| `registered` | `bool` | 1 byte  | 
+| | | 39 bytes |
+
+- `codecID` is the codec version used to serialize the payload, and is hardcoded to `0x0000`
+- `typeID` is the payload type identifier and is `0x00000002` for this message
+- `validationID` identifies the validator for the message
+- `registered` is a boolean representing the status of the `validationID`. If true, the `validationID` corresponds to a validator in the current validator set. If false, the `validationID` does not correspond to a validator in the current validator set, and never will in the future.
+
+#### `SubnetValidatorWeightMessage`
```

Comment:
Why change this from `SubnetValidatorWeightUpdateMessage`? I do believe that the latter makes more sense in regards to `SubnetValidatorRegistrationMessage`.

Reply:
The thinking was that this message format is also used for messages from the P-Chain back to the L1, and the P-Chain is only ever willing to sign for the latest validator weight and nonce, as opposed to signing for each individual weight update.

Reply:
Makes sense!